### ISSUE TITLE
Where can I find instructions to deploy a Django + Whitenoise app using gunicorn?

### ISSUE THREAD

**Issue by subodh-malgonde** [evansd/whitenoise#219]
Hi,

I am using Whitenoise to serve static files in my Django app. I am **NOT** using Nginx. I plan to use a CDN like cloudfront in the future.

I have been looking for deployment instructions, which handle these questions:

1. Since I am not using nginx, I plan to bind gunicorn directly to port 80. This results in an error.
2. How to handle SSL certificate stuff? Typically this is handled by servers like Nginx.

I understand that these are beyond the scope of Whitenoise documentation, and this is not exactly a Whitenoise "issue". However I have been trying to figure this out for 2 days now. 

It would be great if someone could point to instructions to do this. 

**EDIT**: I am deploying on a Ubuntu 18.04 on Google Compute Engine. 

**EDIT 2**: There are many tutorials about deploying with Whitenoise on Heroku. But none for deploying on own VM. So I think using nginx will just make my life a lot easier. Now the question is, should I use Whitenoise while I am using nginx as the reverse proxy?

**Comment by subodh-malgonde** [comment]
TL;DR
---
I used nginx as the http server. I removed the configuration associated with static files in nginx, so the static file requests are passed to the wsgi layer (gunicorn) and are handled by Whitenoise. So you can follow any 'nginx + gunicorn + django' deployment instructions/tutorials, which are easily available with a simple google search.

This post cleared it up for me: [Deploying your Django static files to AWS (Part 2)](https://www.dedi.co/blog/entries/2019/01/02/deploying-your-django-static-files-aws-partnbsp2).

Long Answer
---
As mentioned before there are many tutorials about deploying Django + Whitenoise applications on Heroku. So I had posted this [question on StackOverflow](https://stackoverflow.com/questions/55934127/how-to-deploy-a-django-whitenoise-application-using-gunicorn) and someone pointed out that:

>Heroku, which has its own proxy layer in the front end, and is therefore not at all relevant.

Without verifying this statement, I thought this must be true. gunicorn is not a full fledged webserver. In fact gunicorn creators strongly recommend using it behind a proxy server (Eg. Nginx). [See docs](http://docs.gunicorn.org/en/stable/deploy.html). 

I was confused because I always thought of Nginx as just a reverse proxy. Functioning as a **reverse proxy** for static assets is *just one of the functions* of nginx. It provides a lot more features like buffering slow clients, which gunicorn does not, which helps prevent denial-of-service attacks.

I knew this already. It would have been foolish to not use nginx or any other webserver.

Whitenoise is just there to set proper caching headers for static files and enable compression using gzip/brotli. **If you are note deploying on Heroku you will still need a web server like Nginx**. So you can follow any 'nginx + gunicorn + django' deployment instructions/tutorials, which are easily available with a simple google search. One of which is [Deploying your Django static files to AWS (Part 2)](https://www.dedi.co/blog/entries/2019/01/02/deploying-your-django-static-files-aws-partnbsp2), which helped me get this issue sorted out.

**Comment by franciscojavierarceo** [comment]
@subodh-malgonde I wasn't able to get WhiteNoise working on GCP's CDN, unfortunately. Would you be able provide a little more information about what you did? I followed the blog posts you suggested.

**Comment by subodh-malgonde** [comment]
@franciscojavierarceo You don't have to do anything special. 

1. You just have to configure the CDN to cache static assets (css, js and images)
2. Ensure that your static file requests are passed to the Django server. Usually people use Nginx infront the django server and redirect static file requests to the [STATIC_ROOT](https://docs.djangoproject.com/en/3.0/ref/settings/#std:setting-STATIC_ROOT). However, **you don't have to do this** when using Whitenoise.

### How does Whitenoise work?
When your static file is first requested, it will be sent to the Django server. After this the file will be cached by your CDN and all subsequest requests (across different users) will be served by the CDN. They won't hit your server. 

Whitenoise creates versions of your static files. Eg. If you have a file named `image.png` then when you do `python manage.py collectstatic`, whitenoise will create a file `image.49ec9402.png`. When the file is first requested Whitenoise will add caching headers which will instruct the CDN to cache the file indefinitely. If you edit the file, Whitenoise will change the version of the file, so you don't have worry about stale assets.

**Comment by franciscojavierarceo** [comment]
Thanks so much @subodh-malgonde for following up with me! I have my application currently deployed with Heroku using WhiteNoise, so, fortunately, I am familiar with how it operates.

That said, I'm trying to migrate to Google App Engine and the configuring WhiteNoise is my last step. I have Google CDN set up to a fixed IP address and I have my storage bucket enabled properly but WhiteNoise doesn't seem to actually be cacheing my files. I confirmed this by testing both deployments on https://gtmetrix.com/.

I tried using both the Google storage bucket URL and the fixed IP Address but neither of them enabled WhiteNoise to work correctly. 

Not sure what else to try...thanks so much for your help!

**Comment by evansd** [comment]
@franciscojavierarceo I don't know any of the details of your setup, but the fact that you mention "storage bucket" suggests that you're not configuring Whitenoise in the usual way.

I'd say the first step is to forget about the CDN and make sure Whitenoise is serving your files with the correct caching headers (just use web inspector to see if it's setting an expiry date far into the future).
